scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1493 @@
+            remainder = perm[np.array(remainder).astype(int)]
+
+        sequence += remainder * b2r
+        b2r /= base
+        quotient = (quotient - remainder) / base
+
+    return sequence
+
+
+class QMCEngine(ABC):
+    """A generic Quasi-Monte Carlo sampler class meant for subclassing.
+
+    QMCEngine is a base class to construct specific Quasi-Monte Carlo sampler.
+    It cannot be used directly as a sampler.
+
+    Parameters
+    ----------
+    dim : int
+        Dimension of the parameter space.
+    seed : {int or `numpy.random.RandomState` instance}, optional
+        If `seed` is not specified the `numpy.random.RandomState`
+        singleton is used.
+        If `seed` is an int, a new ``RandomState`` instance is used,
+        seeded with `seed`.
+        If `seed` is already a ``RandomState`` instance, then that
+        instance is used.
+
+    Notes
+    -----
+    By convention samples are distributed over the half-open interval
+    ``[0, 1)``. Instances of the class can access the attributes: ``dim`` for
+    the dimension; and ``rng`` for the random number generator (used for the
+    ``seed``).
+
+    **Subclassing**
+
+    When subclassing `QMCEngine` to create a new sampler,  ``__init__`` and
+    ``random`` must be redefined.
+
+    * ``__init__(dim, seed=None)``: at least fix the dimension. If the sampler
+      does not take advantage of a ``seed`` (deterministic methods like
+      Halton), this parameter can be omitted.
+    * ``random(n_samples)``: draw ``n_samples`` from the engine.
+
+    Optionally, 2 other methods can be overwritten by subclasses:
+
+    * ``reset``: Reset the engine to it's original state.
+    * ``fast_forward``: It should be used as a way to fast-forward a sequence
+      to a further state. If the sequence is deterministic (like Halton
+      sequence), then ``fast_forward(n)`` is skipping the ``n`` first draw.
+
+    Examples
+    --------
+    To create a random sampler based on ``np.random.random``, we would do the
+    following:
+
+    >>> from scipy.stats import qmc
+    >>> class RandomEngine(qmc.QMCEngine):
+    ...     def __init__(self, dim, seed):
+    ...         super().__init__(dim=dim, seed=seed)
+    ...
+    ...
+    ...     def random(self, n_samples=1):
+    ...         return self.rng.random((n_samples, self.dim))
+    ...
+
+    We subclass `QMCEngine` by defining the sampling strategy we want to use.
+    And we can create an instance to sample from.
+
+    >>> engine = RandomEngine(2, seed=12345)
+    >>> engine.random(5)
+    array([[0.92961609, 0.31637555],
+           [0.18391881, 0.20456028],
+           [0.56772503, 0.5955447 ],
+           [0.96451452, 0.6531771 ],
+           [0.74890664, 0.65356987]])
+
+    """
+
+    @abstractmethod
+    def __init__(self, dim, seed=None):
+        self.dim = dim
+        self.rng = check_random_state(seed)
+        self.num_generated = 0
+
+    @abstractmethod
+    def random(self, n_samples=1):
```

Comment:
IMO this is just more explicit and looks more pythonic. I am not a fan of having one letter variables. I will change it if you want.

Reply:
Either is fine with me. I just think it should be consistent between `fast_forward` and `random`. If you go with something like `n_samples` (or `size`, as in `np.random` - although that allows tuples) for both, also reconsider the single letter `m` in `random_base2`. On the other hand, there is plenty of precedent for single-letter variables in scipy.stats, so it would not be out of place.

Reply:
Ok I will change all to `n`, `d` and `m` if you say it's fine like that.